README
email2trac - utilities
----------------------

This is a release of the SARA package email2trac that contains utilities that 
we use to convert emails to trac tickets. The initial setup was made by 
Daniel Lundin from Edgewall Software. SARA has extend the initial setup,  
with the following extensions:

  - HTML message conversion
  - Attachments
  - Tickets can be updated via email
  - Use command-line options
  - Configuration file to control the behavior.
  - Unicode support
  - SPAM detection
  - Workflow support
  - FullBlogPlugin support
  - DiscussionPlugin support 

See INSTALL for the how to setup the utilities

Comments or Suggestions mail them to: 
	email2trac <email2trac@sara.nl>

http://www.sara.nl for more info about SARA.


ChangeLog
2.X.X
  * use self.env.get_read_db for trac version > 0.11. self.env.get_db_cnx is obsolete. TRAC
    api documentation.
    Fixed by: Bas van der Vlies

2.6.2
  * Removed '--Orignal message---' from strip_signature_regex. It is not a 
    signature but a quote. Thanks to Walter Lioen (SARA)
    Fixed by: Bas van der Vlies

  * Added support fot trac 1.0, closes #302, #303
    Reported by: thomas dot schabetsberger add webtek dot at
    Fixed by: Bas van der Vlies

2.6.0 (Dedicated to Ian Verhey (RIP) )

  * License changed from GNU Public License to Apache License, Version 2.0.
    Now the software can be included in Apache projects like bloodhound,
    closes #292

    Requested by:  gary dot martin add wandisco dot com

  * Catch an error when no valid trac installation can be found, 
    closes #294
    Reported by: m dot frigui add cacom dot fr
    Fixed by: Bas van der Vlies

  * Ticket update honors the reply_all setting and sets the ticket cc
    field if there are cc-addresses in the email, closes #293

    Reported by: Sergey V.Levin <slevin add adriver dot ru >

    Fixed by: Bas van der Vlies

  * Reporter email address was not skipped and therefor added to cc field
    when reply_all was set.
    Author:  Bas van der Vlies

  * Default setting for resolution can only be set in email2trac.conf. Skip the trac default
    setting, closes #299
    Reported by: ruediger dot kupper add gmail dot com
    Fixed by: Bas van der Vlies

  * Attachments on Windows are not save due the fact that  os.pathconf is not supported. Set
    filename max length to 240 chars on windows, closes #300
    Reported by: Anonymous
    Fixed by: Bas van der Vlies

  * Added a new option 'only_strip_on_update', default value is False. Only call the functions
    strip_quotes and strip_signature for ticket updates, closes #192
    Requested by, Patch:  Konstantin Ryabitsev <icon add mricon dot com>
    Appplied by: Bas van der Vlies

  * Added a new option 'strip_signature_regex', default value: 
      * strip_signature_regex : ^-----Original Message-----$|^-- $
    This will strip signatures form the email messages. People can set their own regular
    expressions, closes #296, #155
    Author: steverweber add gmail dot com
    Applied by: Bas van der Vlies

  * email_quote can now handle regular expresions, default is:
 	email_quote : ^> .*
    We now use the re.match function instead of line.startswith, closes #173
    Suggested/Patch by: samuel add hoffstaetter dot com
    Applied by: Bas van der Vlies

2.5.0 
  * BLOG improvements by Thomas Moschny, closes #287,#235,#175:
   - Tries to use the 'Date:' email header to set the date of the blog
     post or comment.
   - Allows overriding author and categories via subject parameters or
     inline properties (see below).
   - Does not add the 'blog_' prefix to the default (short-) name of a
     blog entry anymore, as it would be rather redundant.
   - Allows selecting a custom short name for the blog entry instead of
     a default date-based short name. The custom short name may not
     contain spaces. (Note that the TracFullBlogPlugin imposes
     additional rules for valid short names.)
   - Bugfix: Fixes crash wile processing mails with attachments (#175).
   - Properly adds attachments to blog entries (#175).

    Author: Thomas Moschny
    Applied by: Bas van der Vlies

  * Added new parameters:
     - white_list_registered_users
     - white_list_file
  
   white_list_registered_users:  
    This will allow registered user to create or update a ticket even
    when they are not added to the 'white_list' in the configuration 
    file.

   white_list_file:
    This will read allowed users from a file. One email address per line.

   Suggested by: randy add psg dot com, closes ticket #288
   Author: Bas van der Vlies

2.4.7 
  * Some blog improvements. Removes 'blog:' from subject line and honors the 
    -n/--dry-run flag, closes #235 
    Author: Thomas Moschny
    Applied by: Bas van der Vlies
  
  * Added Astaro firewall spam handling. Check X-Spam-Flag and X-Spam-Result
    message header flags. see:
     * http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=644916
    Author: W. Martin Borgert" <debacle at debian dot org>
    Applied by: Bas van der Vlies

  * mailto link contained the wrong encoded subject line, charset was not 
    stripped.
    Reported by: Walter Lioen (SARA)
    Fixed by: Bas van der Vlies

2.4.5 
  * Allow regular expressions for recipient_list (multiple to-addresses), 
    white_list and black_list. Each address is checked againt this regular 
    expression, closes #272
    examples:
     * recipient_list : sara.nl (to-address contains sara.nl)
     * recipient_list : ^basv@sara.nl$ (to-address equals basv@sara.nl)
    Reported by: anonymous
    Fixed by: Bas van der Vlies


  * Enhanced run_email2trac to support supplementary groups (initgroups). 
    configure will check if the OS support initgroups function call,
    closes #274
    Author: Dennis McRitchie <dmcr add Princeton dot EDU >
    Applied by: Bas van der Vlies

  * Fixed an error in parsing the From address. Email address was wrong if
    email name contained a charset definition. There error was triggered 
    if 'mailto_link: 1' is set. Resulting in an unicode error.
    Author: Bas van der Vlies

2.4.2 
  * Fixed an error for text/plain attachments that are not specified as 
    inline. These attachments were not attached to the ticket and were
    skipped, closes #270
    Reported by: hju add jochenkuhl dot de
    Fixed by: Bas van der Vlies

  * Use all 'to-field' addresses in message for recipient_list instead of only one, closes #268
    Reported by: anonymous
    Fixed by: Bas van der Vlies

2.4.0 
  * Fixed an error in ticket_update_fields. Always check if given ticket
    field is a valid ticket field for trac. The error could be triggered
    by:
      * ticket_update_by_subject
      * inline_properties


email2trac_test.sh
/usr/bin/python /var/tmp/email2trac-0.3.0/email2trac.py -f /var/tmp/email2trac-0.3.0/email2trac.conf >/var/tmp/out 2>&1


delete_spam.py
#!/usr/bin/python
#
# Copyright (C) 2002
#
# This file is part of the pxeconfig utils
#
# This program is free software; you can redistribute it and/or modify it 
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any
# later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
#
"""
Author: Bas van der Vlies
Date  : 29 September 2205
Desc. : Delete Spam tickets from database. Else we get an lot of 
        tickets

Usage :
	delete_spam [ -f <configfile> ]

defaults:
	configfile = /etc/email2trac.conf

SVN Info:
        $Id: delete_spam.py 1512 2005-10-26 09:20:37Z jaap $
"""
from trac import Environment, Ticket

import os
import sys
import getopt
import shutil
import ConfigParser

def ReadConfig(file):
	"""
	Parse the config file
	"""

	if not os.path.isfile(file):
		print 'File %s does not exists' %file
		sys.exit(1)

	config = ConfigParser.ConfigParser()

	try:
	  config.read(file)
	except ConfigParser.MissingSectionHeaderError,detail:
	  print detail
	  sys.exit(1)

	defaults = config.defaults()
	if not defaults.has_key('project'):
		print 'You have define the location of your trac project, eg:'
		print '\t project: /var/trac/<projectname>'
		sys.exit(1)
	  
	return defaults


def delete_spam(project, debug):
			env = Environment.Environment(project, create=0)
			db = env.get_db_cnx()
			
			cursor = db.cursor()

			# Delete the attachments associated with Spam tickets
			#
			cursor.execute("SELECT id FROM ticket WHERE  component = 'Spam';") 
			while 1:
				row = cursor.fetchone()
				if not row:
					break
				spam_id =  row[0]
				attachment_dir = os.path.join(env.get_attachments_dir(), 'ticket', str(spam_id)) 
				if os.path.exists(attachment_dir):
					if debug:
						print 'delete %s : %s' %(spam_id, attachment_dir)
					shutil.rmtree(attachment_dir)

			cursor.execute("DELETE FROM ticket WHERE  component = 'Spam';") 
			db.commit()

if __name__ == '__main__':
	# Default config file
	#
	configfile = '/etc/email2trac.conf'

	try:
		 opts, args = getopt.getopt(sys.argv[1:], 'hf:', ['help', 'file='])
	except getopt.error,detail:
		print __doc__
		print detail
		sys.exit(1)

	for opt,value in opts:
		if opt in [ '-h', '--help']:
			print __doc__ 
			sys.exit(0) 
		elif opt in ['-f', '--file']:
			configfile = value
	
	settings = ReadConfig(configfile)
	delete_spam(settings['project'], int(settings['debug']))
	print 'Spam is deleted succesfully..' 

# EOB


run_email2trac.c
/*
    run_email2trac.c
    Authors: Bas van der Vlies, Walter de Jong and Michel Jouvin
    SVN Info:
        $Id$

    Only nobody can become the user www-data. Postfix uses this
    user to start an program

    Copyright 2002 SARA

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/
#include "config.h"

#include <sys/types.h>
#include <stdlib.h>
#include <unistd.h>
#include <pwd.h>
#include <sys/stat.h>
#include <string.h>
#include <stdio.h>
#include <limits.h>
#ifdef HAVE_INITGROUPS
#include <grp.h>
#endif

#include "run_email2trac.h"

#ifndef DEBUG
#define DEBUG 0
#endif

void check_username(char *name)
{
  if ( strlen(name) > 30 ) {
    	  if ( DEBUG ) printf("MTA_USERNAME is to large; %s\n", name);
	  exit(-1);
  }
}

int main(int argc, char** argv) {

  int i,j;
  int caller = getuid();
  int status;

  char   **trac_script_args;
  char   *python_egg_cache = NULL;
  struct passwd *TRAC; 
  struct passwd *MTA;
  struct stat script_attrs;
  const char *trac_script = TRAC_SCRIPT_PATH "/" TRAC_SCRIPT_NAME;
 
  /*
  printf("trac_script = %s\n", trac_script);
  */

  /* First copy arguments passed to the wrapper as scripts arguments
     after filtering out some of the possible script options */

  trac_script_args = (char**) malloc((argc+1)*sizeof(char*));
  if (trac_script_args == NULL) {
    if ( DEBUG ) printf("malloc failed\n");
    return 1;
  }
  trac_script_args[0] = TRAC_SCRIPT_NAME;
  for (i=j=1; i<argc; i++) {
    if ( (strcmp(argv[i],"--file") == 0) || 
	 (strcmp(argv[i],"-f") == 0) ) {
      i++;
      continue;
    }
    else if ( (strcmp(argv[i],"--eggcache") == 0) ||
         (strcmp(argv[i],"-e") == 0) ) {
      i++;
      python_egg_cache = argv[i];
      continue;
    }
    
    trac_script_args[j] = argv[i];
    j++;
  }
  trac_script_args[j] = NULL;

  /* Check caller */
  check_username(MTA_USER);
  MTA = getpwnam(MTA_USER);

  if ( MTA == NULL ) {
    if ( DEBUG ) printf("Invalid MTA user (%s)\n", MTA_USER);
    return -3;     /* 253 : MTA user not found */
  }

  if ( caller !=  MTA->pw_uid ) {
    if ( DEBUG ) printf("Invalid caller UID (%d)\n",caller);
    return -2;     /* 254 : Invalid caller */
  }
  
  /* set UID/GID and supplementary groups to be Trac (or apache) user */
  check_username(TRAC_USER);
  if ( TRAC = getpwnam(TRAC_USER) ) {
#ifdef HAVE_INITGROUPS
    if (initgroups(TRAC_USER, TRAC->pw_gid)) {
      if ( DEBUG ) printf("initgroups failed\n");
      return -7;    /* 249 : Can't set supplementary groups */
    }
#endif
    if (setgid(TRAC->pw_gid) || setuid(TRAC->pw_uid)) {
      if ( DEBUG ) printf("setgid or setuid failed\n");
      return -5;   /* 251: Can't set gid or uid */
    }
  } else {
    if ( DEBUG ) printf("Invalid Trac user (%s)\n",TRAC_USER);
    return -6;     /* 250 : Trac user not found */
  }
	 
  /* Check that script exists */
  if ( stat(trac_script,&script_attrs) ) {
    if ( DEBUG ) printf("Script not found (%s)\n",trac_script);
    return -4;    /* 252 : script not found */
  }
 
  /* Set PYTHON_EGG_CACHE env variable if we have been told to do so */
  if ( python_egg_cache != NULL ) {
    setenv("PYTHON_EGG_CACHE",python_egg_cache ,1);
  }

  /* Execute script */
  status = execv(trac_script, trac_script_args);
  
  if ( DEBUG ) printf("Script %s execution failure (error=%d). Check permission and interpreter path.\n",trac_script,status);
  return -1;     /* 255 : should never reach this point */

}

/* EOB */


run_email2trac.h
/*

SVN Info:
    $Id$

Local site configuration
Value defined here are just default values than can be overriden at compile time
See Makefile.

    Copyright 2002 SARA

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/

/* User the MTA is running under */
#ifndef MTA_USER
#define MTA_USER "nobody"
#endif

/* A user with write access to Trac DB */
#ifndef TRAC_USER
#define TRAC_USER "www-data"
#endif

/* email2trac script name and path */
#ifndef TRAC_SCRIPT_NAME
#define TRAC_SCRIPT_NAME "email2trac"
#endif
#ifndef TRAC_SCRIPT_PATH
#define TRAC_SCRIPT_PATH "/usr/bin"
#endif


email2trac.py
#!/usr/bin/python
# Copyright (C) 2002
#
# This file is part of the email2trac utils
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any
# later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
#
"""
emailfilter.py -- Email tickets to Trac.

A simple MTA filter to create Trac tickets from inbound emails.

Copyright 2005, Daniel Lundin <daniel@edgewall.com>
Copyright 2005, Edgewall Software

Changed By: Bas van der Vlies <basv@sara.nl>
Date      : 13 September 2005
Descr.    : Added config file and command line options, spam level
            detection, reply address and mailto option. Unicode support

Changed By: Walter de Jong <walter@sara.nl>
Descr.    : multipart-message code and trac attachments


The scripts reads emails from stdin and inserts directly into a Trac database.
MIME headers are mapped as follows:

	* From:      => Reporter
	             => CC (Optional via reply_address option)
	* Subject:   => Summary
	* Body       => Description
	* Component  => Can be set to SPAM via spam_level option

How to use
----------
 * Create an config file:
	[DEFAULT]                        # REQUIRED
	project      : /data/trac/test   # REQUIRED
	debug        : 1                 # OPTIONAL, if set print some DEBUG info
	spam_level   : 4                 # OPTIONAL, if set check for SPAM mail 
	reply_address: 1                 # OPTIONAL, if set then fill in ticket CC field
        umask        : 022               # OPTIONAL, if set then use this umask for creation of the attachments
	mailto_link  : 1                 # OPTIONAL, if set then [mailto:<CC>] in description 
	trac_version : 0.8               # OPTIONAL, default is 0.9

	[jouvin]                         # OPTIONAL project declaration, if set both fields necessary
	project      : /data/trac/jouvin # use -p|--project jouvin.  
        
 * default config file is : /etc/email2trac.conf

 * Commandline opions:
                -h | --help
                -c <value> | --component=<value>
                -f <config file> | --file=<config file>
                -p <project name> | --project=<project name>

SVN Info:
        $Id: email2trac.py 1523 2005-11-01 09:27:51Z bas $
"""

import os
import sys
import string
import getopt
import stat
import time
import email
import re
import urllib
import unicodedata
import ConfigParser
from email import Header
from stat import *
import mimetypes

trac_default_version = 0.9

class TicketEmailParser(object):
	env = None
	comment = '> '
    
	def __init__(self, env, parameters, version):
		self.env = env

		# Database connection
		#
		self.db = None

		self.VERSION = version
		if self.VERSION > 0.8:
			self.new_ticket = ticket.Ticket
			self.get_config = self.env.config.get
		else:
			self.new_ticket = Ticket.Ticket
			self.get_config = self.env.get_config

		if parameters.has_key('umask'):
			os.umask(int(parameters['umask'], 8))

		if parameters.has_key('debug'):
			self.DEBUG = int(parameters['debug'])
		else:
			self.DEBUG = 0

		if parameters.has_key('reply_address'):
			self.CC = int(parameters['reply_address'])
		else:
			self.CC = 0

		if parameters.has_key('mailto_link'):
			self.MAILTO = int(parameters['mailto_link'])
		else:
			self.MAILTO = 0

		if parameters.has_key('spam_level'):
			self.SPAM_LEVEL = int(parameters['spam_level'])
		else:
			self.SPAM_LEVEL = 0

		if parameters.has_key('email_comment'):
			self.comment = str(parameters['email_comment'])

		if parameters.has_key('email_header'):
			self.EMAIL_HEADER = int(parameters['email_header'])
		else:
			self.EMAIL_HEADER = 0


	# X-Spam-Score: *** (3.255) BAYES_50,DNS_FROM_AHBL_RHSBL,HTML_
	# Note if Spam_level then '*' are included
	def spam(self, message):
		if message.has_key('X-Spam-Score'):
			spam_l = string.split(message['X-Spam-Score'])
			number = spam_l[0].count('*')

			if number >= self.SPAM_LEVEL:
				return number

		return 0

	def to_unicode(self, str):
		"""
		Email has 7 bit ASCII code, convert it to unicode with the charset
		that is encoded in 7-bit ASCII code and encode it as utf-8 so TRAC 
		understands it.
		"""
		results =  Header.decode_header(str)
		str = None
		for text,format in results:
			if format:
				try:
					temp = unicode(text, format)
				except UnicodeError:
					# This always works 
					#
					temp = unicode(text, 'iso-8859-15')
				temp =  temp.encode('utf-8')
			else:
				temp = string.strip(text)

			if str:
				str = '%s %s' %(str, temp)
			else:
				str = temp

		return str

	def debug_attachments(self, message):
		n = 0
		for part in message.walk():
			if part.get_content_maintype() == 'multipart':      # multipart/* is just a container
				print 'TD: multipart container'
				continue

			n = n + 1
			print 'TD: part%d: Content-Type: %s' % (n, part.get_content_type())
			print 'TD: part%d: filename: %s' % (n, part.get_filename())

			if part.is_multipart():
				print 'TD: this part is multipart'
				payload = part.get_payload(decode=1)
				print 'TD: payload:', payload
			else:
				print 'TD: this part is not multipart'

			part_file = '/var/tmp/part%d' % n
			print 'TD: writing part%d (%s)' % (n,part_file)
			fx = open(part_file, 'wb')
			text = part.get_payload(decode=1)
			if not text:
				text = '(None)'
			fx.write(text)
			fx.close()
			try:
				os.chmod(part_file,S_IRWXU|S_IRWXG|S_IRWXO)
			except OSError:
				pass

	def email_header_txt(self, m):
#		if not m['Subject']:
#			subject = '(geen subject)'
#		else:
#			subject = self.to_unicode(m['Subject'])
#
#		head = "'''Subject:''' %s [[BR]]" % subject
#		if m['From'] and len(m['From']) > 0:
#			head = "%s'''From:''' %s [[BR]]" % (head, m['From'])
#		if m['Date'] and len(m['Date']) > 0:
#			head = "%s'''Date:''' %s [[BR]]" %(head, m['Date'])

		str = ''
		if m['To'] and len(m['To']) > 0 and m['To'] != 'hic@sara.nl':
			str = "'''To:''' %s [[BR]]" %(m['To'])
		if m['Cc'] and len(m['Cc']) > 0:
			str = "%s'''Cc:''' %s [[BR]]" % (str, m['Cc'])

		return str

	def parse(self, fp):
		msg = email.message_from_file(fp)
		if not msg:
			return

		if self.DEBUG > 1:	  # save the entire e-mail message text
			msg_file = '/var/tmp/msg.txt' 
			print 'TD: saving email to %s' % msg_file
			fx = open(msg_file, 'wb')
			fx.write('%s' % msg)
			fx.close()
			try:
				os.chmod(msg_file,S_IRWXU|S_IRWXG|S_IRWXO)
			except OSError:
				pass

		self.db = self.env.get_db_cnx()
		tkt = self.new_ticket(self.env)
		tkt['status'] = 'new'

		# Some defaults
		#
		tkt['milestone'] = self.get_config('ticket', 'default_milestone')
		tkt['priority'] = self.get_config('ticket', 'default_priority')
		tkt['severity'] = self.get_config('ticket', 'default_severity')
		tkt['version'] = self.get_config('ticket', 'default_version')

		if not msg['Subject']:
			tkt['summary'] = '(geen subject)'
		else:
			tkt['summary'] = self.to_unicode(msg['Subject'])

		if self.SPAM_LEVEL and self.spam(msg):
			print 'This message is a SPAM. Automatic ticket insertion refused (SPAM level > %d' % self.SPAM_LEVEL
			sys.exit(1)

		if settings.has_key('component'):
			tkt['component'] = settings['component']
		else:
			tkt['component'] = self.get_config('ticket', 'default_component')

		# Get default owner for component
		#
		cursor = self.db.cursor()
		sql = 'SELECT owner FROM component WHERE name=\'%s\'' % tkt['component']
		cursor.execute(sql)
		tkt['owner'] = cursor.fetchone()[0]

		from_str = self.to_unicode(msg['from'])
		tkt['reporter'] = from_str
		if self.CC:
			tkt['cc'] = from_str

# produce e-mail like header
		head = ''
		if self.EMAIL_HEADER > 0:
			head = self.email_header_txt(msg)

		if self.DEBUG > 0:
			self.debug_attachments(msg)

#
#	put the message text in the ticket description
#	message text can be plain text or html or something else
#
		has_description = 0
		for part in msg.walk():
			if part.get_content_maintype() == 'multipart':			# 'multipart/*' is a container for multipart messages
				continue

			if part.get_content_type() == 'text/plain':
				body_text = part.get_payload(decode=1)			# try to decode
				if not body_text:					# decode failed
					body_text = part.get_payload(decode=0)		# do not decode

				tkt['description'] = '\n{{{\n\n%s\n}}}\n' % body_text

			elif part.get_content_type() == 'text/html':
				tkt['description'] = '%s\n\n(see attachment for HTML mail message)\n' % head
				body_text = tkt['description']

			else:
				tkt['description'] = '%s\n\n(see attachment for message)\n' % head
				body_text = tkt['description']

			has_description = 1
			break		# we have the description, so break

		if not has_description:
			tkt['description'] = '%s\n\n(no plain text message, see attachments)' % head
			has_description = 1

		author, email_addr  = email.Utils.parseaddr(msg['from'])
		if self.MAILTO:
			mailto = self.html_mailto_link(author, email_addr, self.to_unicode(msg['subject']), body_text)
			tkt['description'] = '%s\n%s %s' %(head, mailto, tkt['description'])

		if self.VERSION > 0.8:
			tkt['id'] = tkt.insert()
		else:
			tkt['id'] = tkt.insert(self.db)

		#
		# Just how to show to update description
		#
		#tkt['description'] = '\n{{{\n\n Bas is op nieuw bezig\n\n }}}\n' 
		#tkt.save_changes(self.db, author, "Lekker bezig")
		#
		self.attachments(msg, tkt, author)


	def mail_line(self, str):
		return '%s %s' % (self.comment, str)


	def html_mailto_link(self, author, mail_addr, subject, body):
		if not author:
			author = mail_addr
		else:	
			author = self.to_unicode(author)

		# Must find a fix
		#
		#arr = string.split(body, '\n')
		#arr = map(self.mail_line, arr)

		#body = string.join(arr, '\n')
		#body = '%s wrote:\n%s' %(author, body)

		# Obsolete for reference
		#
		#body = self.to_unicode(body)
		#body = urllib.quote(body)
		#body = Header.encode(body)
		#

		# Temporary fix
		body = '> Type your reply'
		str = 'mailto:%s?subject=%s&body=%s' % (urllib.quote(mail_addr), urllib.quote('Re: %s' % subject), urllib.quote(body))
		str = '\n{{{\n#!html\n<a href="%s">Reply to: %s</a>\n}}}\n' %(str, author)

		return str

	def attachments(self, message, ticket, user):
		'''save any attachments as file in the ticket's directory'''

		count = 0
		first = 0
		for part in message.walk():
			if part.get_content_maintype() == 'multipart':		# multipart/* is just a container
				continue

			if not first:										# first content is the message
				first = 1
				if part.get_content_type() == 'text/plain':		# if first is text, is was already put in the description
					continue

			filename = part.get_filename() 
			if not filename:
				count = count + 1
				filename = 'part%04d' % count

				ext = mimetypes.guess_extension(part.get_type())
				if not ext:
					ext = '.bin'

				filename = '%s%s' % (filename, ext)
			else:
				filename = self.to_unicode(filename)

			if '/' in filename:
				filename = os.path.basename(filename)

			url_filename = urllib.quote(filename)

			if self.VERSION > 0.8:
				tmpfile = '/tmp/email2trac-ticket%sattachment' % str(ticket['id'])
			else:
				dir = os.path.join(self.env.get_attachments_dir(), 'ticket', 
				                    	urllib.quote(str(ticket['id'])))
				if not os.path.exists(dir):
					mkdir_p(dir, 0755)

				tmpfile = os.path.join(dir, url_filename)

			f = open(tmpfile, 'wb')
			text = part.get_payload(decode=1)
			if not text:
				text = '(None)'
			f.write(text)

			# get the filesize
			#
			stats = os.lstat(tmpfile) 
			filesize = stats[stat.ST_SIZE]

			# Insert the attachment it differs for the different TRAC versions
			#
			if self.VERSION > 0.8:
				att = attachment.Attachment(self.env,'ticket',ticket['id'])
				att.insert(url_filename,f,filesize)
				f.close()
			else:
				cursor = self.db.cursor()
				cursor.execute('INSERT INTO attachment VALUES("%s","%s","%s",%d,%d,"%s","%s","%s")' 
					%('ticket', urllib.quote(str(ticket['id'])), filename + '?format=raw', filesize, 
					  int(time.time()),'', user, 'e-mail') ) 
				self.db.commit()


def mkdir_p(dir, mode):
	'''do a mkdir -p'''

	arr = string.split(dir, '/')
	path = ''
	for part in arr:
		path = '%s/%s' % (path, part)
		try:
			stats = os.stat(path)
		except OSError:
			os.mkdir(path, mode)


def ReadConfig(file, name):
	"""
	Parse the config file
	"""

	if not os.path.isfile(file):
		print 'File %s does not exists' %file
		sys.exit(1)

	config = ConfigParser.ConfigParser()
	try:
		config.read(file)
	except ConfigParser.MissingSectionHeaderError,detail:
		print detail
	  	sys.exit(1)


  	# Use given project name else use defaults
  	#
	if name:
		if not config.has_section(name):
			print "Not an valid project name: %s" %name
			print "Valid names: %s" %config.sections()
			sys.exit(1)

		project =  dict()
		for option in  config.options(name):
			project[option] = config.get(name, option) 

	else:
		project = config.defaults()

	return project

if __name__ == '__main__':
	# Default config file
	#
	configfile = '/etc/email2trac.conf'
	# configfile = './email2trac.conf'
	project = ''
	component = ''
	
	try:
		opts, args = getopt.getopt(sys.argv[1:], 'chf:p:', ['component=','help', 'file=', 'project='])
	except getopt.error,detail:
		print __doc__
		print detail
		sys.exit(1)

	project_name = None
	for opt,value in opts:
		if opt in [ '-h', '--help']:
			print __doc__
			sys.exit(0)
		elif opt in ['-c', '--component']:
			component = value
		elif opt in ['-f', '--file']:
			configfile = value
		elif opt in ['-p', '--project']:
			project_name = value

	settings = ReadConfig(configfile, project_name)
	if not settings.has_key('project'):
		print __doc__
		print 'No project defined in config file, eg:\n\t project: /data/trac/bas'
		sys.exit(1)

	if component:
		settings['component'] = component

	if settings.has_key('trac_version'):
		version = float(settings['trac_version'])
	else:
		version = trac_default_version

	#debug HvB
	#print settings

	if version > 0.8:
		from trac import attachment, config, env, ticket
		env = env.Environment(settings['project'], create=0)
		ticket_mod = ticket
	else:
		from trac import Environment, Ticket
		env = Environment.Environment(settings['project'], create=0)
		ticket_mod = Ticket
		
	tktparser = TicketEmailParser(env, settings, version)
	tktparser.parse(sys.stdin)

# EOB
